grype/matcher/java/matcher.go

```diff
@@ -9,7 +16,104 @@ import (
 	syftPkg "github.com/anchore/syft/syft/pkg"
 )
 
+const (
+	sha1Query = `1:"%s"`
+)
+
 type Matcher struct {
+	SearchMavenUpstream bool
+	MavenSearcher
+}
+
+// MavenSearcher is the interface that wraps the GetMavenPackageBySha method.
+type MavenSearcher interface {
+	// GetMavenPackageBySha provides an interface for building a package from maven data based on a sha1 digest
+	GetMavenPackageBySha(string) (*pkg.Package, error)
+}
+
+// mavenSearch implements the MavenSearcher interface
+type mavenSearch struct {
+	client  *http.Client
+	baseURL string
+}
+
+type mavenAPIResponse struct {
+	Response struct {
+		NumFound int `json:"numFound"`
+		Docs     []struct {
+			ID           string `json:"id"`
+			GroupID      string `json:"g"`
+			ArtifactID   string `json:"a"`
+			Version      string `json:"v"`
+			P            string `json:"p"`
+			VersionCount int    `json:"versionCount"`
+		} `json:"docs"`
+	} `json:"response"`
+}
+
+func (ms *mavenSearch) GetMavenPackageBySha(sha1 string) (*pkg.Package, error) {
+	req, err := http.NewRequest(http.MethodGet, ms.baseURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to initialize HTTP client: %w", err)
+	}
+
+	q := req.URL.Query()
+	q.Set("q", fmt.Sprintf(sha1Query, sha1))
+	q.Set("rows", "1")
+	q.Set("wt", "json")
+	req.URL.RawQuery = q.Encode()
+
+	resp, err := ms.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("sha1 search error: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("status %s from %s", resp.Status, req.URL.String())
+	}
+
+	var res mavenAPIResponse
+	if err = json.NewDecoder(resp.Body).Decode(&res); err != nil {
+		return nil, fmt.Errorf("json decode error: %w", err)
+	}
+
+	if len(res.Response.Docs) == 0 {
+		return nil, fmt.Errorf("digest %s: %w", sha1, errors.New("no artifact found"))
+	}
+
+	// artifacts might have the same SHA-1 digests.
```

Comment:
Right, this is a good point; would it make sense to somehow check if this matches the original package name in some way?